pkg/specification/macspec/specification.go

```diff
@@ -0,0 +1,242 @@
+// limitations under the License.
+
+// Package macspec provides access to LoRaWAN MAC specification-specific settings.
+package macspec
+
+import (
+	"fmt"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/crypto"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+// macVersion returns the MACVersion as an integer.
+func macVersion(v ttnpb.MACVersion) int {
+	switch v {
+	case ttnpb.MACVersion_MAC_V1_0:
+		return 100
+	case ttnpb.MACVersion_MAC_V1_0_1:
+		return 101
+	case ttnpb.MACVersion_MAC_V1_0_2:
+		return 102
+	case ttnpb.MACVersion_MAC_V1_0_3:
+		return 103
+	case ttnpb.MACVersion_MAC_V1_0_4:
+		return 104
+	case ttnpb.MACVersion_MAC_V1_1:
+		return 110
+	default:
+		panic(fmt.Errorf("missed %q in macVersion", v))
+	}
+}
+
+// compareMACVersion compares MACVersions lhs to rhs:
+// -1 == lhs is less than rhs
+// 0 == lhs is equal to rhs
+// 1 == lhs is greater than rhs
+func compareMACVersion(lhs, rhs ttnpb.MACVersion) int {
+	switch r := macVersion(lhs) - macVersion(rhs); {
+	case r < 0:
+		return -1
+	case r == 0:
+		return 0
+	case r > 0:
+		return 1
+	default:
+		panic("unreachable")
+	}
+}
+
+// EncryptFOpts reports whether v requires MAC commands in FOpts to be encrypted.
+func EncryptFOpts(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
+
+// HasMaxFCntGap reports whether v defines a MaxFCntGap.
+func HasMaxFCntGap(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) < 0
+}
+
+// HasNoChangeTXPowerIndex reports whether v defines a no-change TxPowerIndex value.
+func HasNoChangeTXPowerIndex(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0
+}
+
+// HasNoChangeDataRateIndex reports whether v defines a no-change DataRateIndex value.
+func HasNoChangeDataRateIndex(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0
+}
+
+// IgnoreUplinksExceedingLengthLimit reports whether v requires Network Server to
+// silently drop uplinks exceeding selected data rate payload length limits.
+func IgnoreUplinksExceedingLengthLimit(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0 && compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// IncrementDevNonce reports whether v defines DevNonce as an incrementing counter.
+func IncrementDevNonce(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_0_4) >= 0
+}
+
+// UseNwkKey reports whether v uses a root NwkKey.
+func UseNwkKey(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) >= 0
+}
+
+// UseLegacyMIC reports whether v uses legacy MIC computation algorithm.
+func UseLegacyMIC(v ttnpb.MACVersion) bool {
+	return compareMACVersion(v, ttnpb.MACVersion_MAC_V1_1) < 0
+}
+
+// UseSharedFCntDown reports whether v uses the same frame counter
+// for both network and application downlinks.
+func UseSharedFCntDown(v ttnpb.MACVersion) bool {
```

Comment:
`UseSingleFCntDown` ?

Reply:
I named it 'shared' since this makes it clear that both entities (application layer and network layer) basically fight/negotiate which value to use. The fact that the frame counter is single could also be understood as only one layer needs to manage it.